hydromt/raster.py

```diff
@@ -2233,6 +2234,243 @@ def tile_window(shape, px):
+
+        def create_folder(path):
+            if not os.path.exists(path):
+                os.makedirs(path)
+
+        def transform_res(dres, transformer):
+            return transformer.transform(0, dres)[0]
+
+        transformer = pyproj.Transformer.from_crs(self._obj.raster.crs.to_epsg(), 3857)
+
+        if bbox:
+            minx, miny, maxx, maxy = bbox
+        else:
+            minx, miny, maxx, maxy = obj.raster.transform_bounds("EPSG:3857")
+
+        minx, miny = map(
+            max,
+            zip((minx, miny), [-20037508.34] * 2),
+        )
+        maxx, maxy = map(
+            min,
+            zip((maxx, maxy), [20037508.34] * 2),
+        )
+
+        dres = abs(self._obj.raster.res[0])
+        dres = transform_res(dres, transformer)
+        nzl = int(np.ceil((np.log10((20037508.34 * 2) / (dres * 256)) / np.log10(2))))
+
+        if max(zoom_levels) > nzl:
+            raise OSError("")
+
+        def tile_window(zl, minx, miny, maxx, maxy):
+            # Origin displacement
+            odx = np.floor(abs(-20037508.34 - minx) / dx)
+            ody = np.floor(abs(20037508.34 - maxy) / dx)
+
+            # Set the new origin
+            minx = -20037508.34 + odx * dx
+            maxy = 20037508.34 - ody * dx
+
+            # Create window generator
+            lu = product(
+                np.arange(round(minx, 2), round(maxx, 2), dx),
+                np.arange(round(maxy, 2), round(miny, 2), -dx),
+            )
+            for l, u in lu:
+                col = int(round(odx + (l - minx) / dx, 2))
+                row = int(round(ody + (maxy - u) / dx, 2))
+                bounds = [
+                    l,
+                    u - dx,
+                    l + dx,
+                    u,
+                ]
+                yield Affine(px, 0, l, 0, -px, u), bounds, col, row
+
+        zoom_levels.sort()
+        if zoom_levels[-1] > min_lvl:
+            min_lvl = zoom_levels[-1]
+        zoom_levels = list(range(zoom_levels[0], min_lvl + 1, 1))
+        zoom_levels.reverse()
+
+        create_folder(root)
+
+        vrt_fn = None
+        nodata = self.nodata
+        zls = {}
+        for zl in zoom_levels:
+            dx = (20037508.34 * 2) / (2**zl)
+            px = dx / px_size
+
+            # read data from previous zoomlevel
+            if vrt_fn is not None:
+                obj = xr.open_dataarray(vrt_fn, engine="rasterio").squeeze(
+                    "band", drop=True
+                )
+
+            obj_bounds = obj.raster.bounds
+            obj_res = obj.raster.res[0]
+            obj_shape = obj.raster.shape
+
+            t_r = pyproj.Transformer.from_crs(
+                3857, obj.raster.crs.to_epsg(), always_xy=True
+            )
+
+            sd = join(root, f"{zl}")
+            create_folder(sd)
+            txt_path = join(sd, "filelist.txt")
+            file = open(txt_path, "w")
+
+            for transform, bounds, col, row in tile_window(zl, minx, miny, maxx, maxy):
```

Comment:
Would it make sense to make a separate function for the part within this loop? In that case python takes care of your garbage collection much better and you don't have to work with del and gc.collect.

Reply:
I manually call the garbage collection between iterations to clear some memory that is not automatically cleared (in order to avoid creeping memory), a seperate function would not solve that.